### DISCUSSION TITLE
IP packet exceeding interface MTU, Adjusting ipsec MTU

### DISCUSSION THREAD

**Discussion by jxcang** [strongswan/strongswan#1357] [Q&A]
Set up an ipsec tunnel between a headless computer running Alpine Linux as router and Cisco RV260 VPN router.
Alpine Linux is a minimal setup, installed iproute2, iptables, strongswan (compiled from source, no rc-service files, enabled bypass-lan), busybox-extras (udhcpd, syslog).

IPsec establishes and runs with no problem, except certain websites didn't load. All interface MTUs at default setting of 1500. Googling reveals a possible MTU issue.

1) Adjusting client device Ethernet/wireless MTU to 1380 solves the issue for the ones that could be adjusted. IPhone iOS cannot be adjusted.
2) Restoring the interface MTU to default and set iptables, FORWARD chain, -p tcp -m tcp --tcp-flags SYN,RST SYN -j TCPMSS --set-mss 800, worked for all the devices with default interface MTU.

Is there a way to not use the iptables set-mss and instead use IP fragmentation to break up the packets after coming from the LAN port at MTU 1500, and before hitting xfrm policies and being re-packaged in another IP header before sending out on WAN port?

I don't know whether the following setup is a good idea or over-engineering. I am accused by my colleagues to be a frequent over-engineer. Keep LAN port in main network namespace (ns) and move WAN port to new ns. Connect main and new ns'es with vEthernet. Run strongswan in new ns. Bridge LAN port with vEthernet in main ns. Keep MTU 1500 on LAN port, and set MTU 1380 on vEthernet. Would the Linux kernel start fragmenting the packet coming from LAN port before forwarding to vEthernet?

**Comment by tobiasbrunner** [comment]
Not sure if your elaborate setup would work, but I think XFRM interfaces could be used to enforce fragmentation before tunneling (in particular when working with UDP). In some kernel versions, setting the MSS/MTU on the route can also work. But modifying the MSS via firewall is usually the simplest way to avoid MTU issues (going as low as 800 is usually not required, though - the overhead depends on the algorithms etc.). See [here](https://docs.strongswan.org/docs/5.9/howtos/forwarding.html#_mtumss_issues) for more info.

**Reply by jxcang** [reply]
I am confused about the ip xfrm implementation. I don't see a xfrm interface in ip link, and I don't believe there is an ip link with type xfrm. Is there a way for me to adjust the MTU for xfrm interface through ip xfrm command suite?
I can try to set MTU via ip route. I guess I will adjust the MTU on the default route from WAN port. Do I need to use mangle table to clear the DF bit if set? Or enable ICMP type 3 from LAN port to local subnet clients for PMTUD?
The thing with TCP mss is that UDP is not taken care of, so QUIC would still glitch if exceeding 1500 MTU after going through ipsec.

**Reply by Thermi** [reply]
>  and I don't believe there is an ip link with type xfrm

There is.

E.g.

`ip link help`
```
[...]
TYPE := { amt | bareudp | bond | bond_slave | bridge | bridge_slave |
[...]
          xfrm | virt_wifi }
```
> Is there a way for me to adjust the MTU for xfrm interface through ip xfrm command suite?

No, the Interface MTU is changed by using ip link.

> I can try to set MTU via ip route.

Can work, but not in all cases. Some applications using, for example, sctp are known to behave weirdly and ignore the route MTU. I've seen it with some proprietary stuff.

> uess I will adjust the MTU on the default route from WAN port. 

You can do it via the mtu confg value of the kernel-netlink plugin.

>  Do I need to use mangle table to clear the DF bit if set?

iptables doesn't support that (weirdly).

> Or enable ICMP type 3 from LAN port to local subnet clients for PMTUD?

That's the default behavior anyway unless you drop ICMP packets (which you shouldn't of course)

> The thing with TCP mss is that UDP is not taken care of, so QUIC would still glitch if exceeding 1500 MTU after going through ipsec.

Yep.

**Reply by jxcang** [reply]
I am using iproute2 4.19.0 and that's out of date. I was wondering why I don't have ip link help xfrm option, yet ip xfrm works. The Alpine Linux image was old too, I set it up a few years ago and finally got some time to build a VPN router out of it as the first project.

Just did apk update to upgrade iproute2 to 6.0.0, and now I can see ip link type with xfrm.

With the CHILD_SA connected, I don't see an ip type xfrm interface being created. I am also using bypass_lan plugin for local LAN routing and remote_ts set to 0.0.0.0/0. Does this I mean have to switch from bypass_lan plugin to xfrm route based VPN?

Note that this is a site to site config scenario. I have control on the remote side using strongswan, but does not have too much control on the local side using Cisco router.

**Reply by Thermi** [reply]
You need to create and configure them manually or via a script. strongSwan doesn't do it for you.

**Reply by jxcang** [reply]
Sorry, I might need some guidance on using xfrm interface. I am reading on the linked page that bypass_lan is not compatible with xfrm route based VPN. So it is an either/or setup from my understanding. If I want to pursue ip xfrm interface MTU, then I need to use xfrm for the LAN bypass as well?

**Reply by Thermi** [reply]
bypass_lan is not any kind of forwarding implementation or something like that. It just installs bypass policies for local LAN traffic to save you from having to configure new passthrough policies any time your laptop roams to a new network.

You don't need to use bypass_lan if you exclusively use route based VPNs.

**Reply by jxcang** [reply]
I am not saying bypass_lan is an equivalent to bridge or alike. I have bypass_lan enabled because the Alpine Linux is acting as a VPN gateway and is routing traffic from LAN to WAN (ipsec tunnel to primary site).

Once I migrate to route based VPN, then I won't need bypas_lan anymore, presumably.
Any guidance as to how to configure the route based VPN to send all traffic except the local subnet over the VPN? I have the remote_ts set to 0.0.0.0/0. The documentation sounds like route based VPN would defeat the purpose of remote_ts 0.0.0.0/0.

Sorry, I may have not understood all items in the documentation as I have a full time job not related to computers at all.

Another thing, I don't have /usr/local/libexec/ipsec/xfrmi utility for some reason. Is this needs to be enabled using a ./configure flag?

**Comment by kohnish** [comment]
> IPhone iOS cannot be adjusted.  
  
Assuming it's about IKEv2, it supports changing it from the default (1280) https://developer.apple.com/documentation/devicemanagement/vpn/ikev2-data.dictionary
Maybe not supported via GUI but 1280 should be good for most setup. 

For strongswan client, setting mtu in /etc/strongswan.d/charon/kernel-netlink.conf works out for me. No traditonal tunnel interface needed just to set mtu there. I don't know if it works on the server side though.  

On my server, I have the following in my mangle table. It limits the alteration to ipsec traffic only. 
```
-A FORWARD -p tcp -m policy --dir in --pol ipsec -m tcp --tcp-flags SYN,RST SYN -m tcpmss --mss 1361:1536 -j TCPMSS --set-mss 1360
-A FORWARD -p tcp -m policy --dir out --pol ipsec -m tcp --tcp-flags SYN,RST SYN -m tcpmss --mss 1361:1536 -j TCPMSS --set-mss 1360
```
